### PR TITLE
More Python Machines

### PR DESCRIPTION
**Pull request by gcarleo**
This PR ports the all-things pure-state RBMs to python 

- [x] RbmSpin
- [x] RbmSpinSymm
- [x] RbmSpinReal 
- [x] RbmSpinPhase 
- [x] RbmMultiVal
- [x] Jastrow
- [x] JastrowSymm  

Important change is that most things are implemented in a single class, instead of doing multiple classes as in the old c++ code. For example, ```RbmSpinReal ``` is just a short-cut for 

```python
RbmSpin(...,dtype=float) 
```

In turn this means that we have some functionalities that weren't available before, for example we can have real-valued ```RbmMultiVal``` just changing dtype in its constructor. 

Symmetries in the RBM yet are implemented with a simple logic, i.e. ```RbmSpinSymm```, is just a short-cut for 

```python
RbmSpin(...,symmetry=True)
```

Notice that now we can pass an arbitrary automorphism vector to symmetry (instead of just passing True, so that a graph is not explicitly required in the incoming hilbert space). 

This immediately allows new machines that weren't available before. Most notably, we now have a nice ```RbmMultiVal``` with symmetries, that is pretty handy/very accurate for the translation-invariant Bose Hubbard model. 
I have update the example in ```BoseHubbard1d/bosehubbard1d.py``` accordingly. 
@VolodyaCO this might be of interest for you if you are still working on the Bose Hubbard (also see #328.

### CODE REVIEW

**Comment by gcarleo** [comment]
While implementing ```RbmSpinSymm``` I faced a possible limitation of the ```state_dict``` approach. 
Basically, I couldn't find a way to work with state_dict and with a view of the parameters rather than working with copies. 

The main issue is that when using symmetries your parameters are a matrix of weights (say, Ws) that maps onto a larger matrix (say W) that is actually used internally to compute log_val etc. 

Now, with numpy it seems impossible to assign W a **view** of Ws, because W=Ws[advanced_index], and this returns a copy, as per numpy policy (it makes sense as a behavior).  Is there a way around this? If not, I cannot see many alternatives but working with copies when assigning parameters... 

@femtobit @twesterhout what do you think?

**Comment by gcarleo** [comment]
@femtobit you mean in numpy-based machines like the RBM? Yeah, I think it should be possible to use lower precision, it might need a few changes though...

**Comment by VolodyaCO** [comment]
@gcarleo Yes, this is useful for me, although I have already intensively tested the BH parameter space with the old RBM MultiVal... so I hope all that work is valuable (probably better results can be achieved with the new symmetric RBM), as I have already finished writing a manuscript (don't know if you want to take a look at it).

File: netket/machine/abstract_machine.py [hunk 1]
```diff
@@ -159,3 +159,7 @@ def save(self, file):
 
     def load(self, file):
         self.parameters = _np.load(file, allow_pickle=False)
+
+    @property
+    def n_visible(self):
+        return self.hilbert.size
```
Comment (PhilipVinc) [inline netket/machine/abstract_machine.py:165]: Shouldn't we change the name of this to `input_size` or something like that?
The name `n_visible` is really something that only makes sense for RBMs...

Comment (gcarleo) [reply netket/machine/abstract_machine.py:165]: yes, this was mostly for backward compatibility

**Comment by gcarleo** [comment]
I would merge this now, since there will be some more work coming depending on this

**Comment by femtobit** [comment]
Fine with me. I'll open an issue or make a new PR if I notice something regarding the machines later on.

**Comment by gcarleo** [comment]
> Fine with me. I'll open an issue or make a new PR if I notice something regarding the machines later on.

yes thanks... ! there is some new very interesting stuff with jax machines that I am doing, more soon...

**Pull request merged**